modules/ml/src/main/java/org/apache/ignite/ml/knn/utils/BallTreeSpatialIndex.java

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.knn.utils;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import org.apache.ignite.ml.math.distances.DistanceMeasure;
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+import org.apache.ignite.ml.math.primitives.vector.VectorUtils;
+import org.apache.ignite.ml.structures.LabeledVector;
+
+import static org.apache.ignite.ml.knn.utils.PointWithDistanceUtil.transfomToListOrdered;
+import static org.apache.ignite.ml.knn.utils.PointWithDistanceUtil.tryToAddIntoHeap;
+
+/**
+ * Ball tree based implementation of {@link SpatialIndex}. Asymptotic runtime complexity of finding {@code k} closest
+ * elements is {@code O(log(n)*k)}, but it degrades on high dimensional data.
+ *
+ * @param <L> Label type.
+ */
+public class BallTreeSpatialIndex<L> implements SpatialIndex<L> {
+    /** Number of points in a leaf. */
+    private static final int MAX_LEAF_SIZE = 42;
```

Comment:
it can be a set bu a user of algorithm

Reply:
User experience is our top priority, so I don't this that it makes sense to inflate the number of parameters in the user interface. Especially in case when it doesn't affect results, only performance (line +- 1%).